File: actuators/magnetorquer.py
import numpy as np
from world.math.quaternions import crossproduct, quatrotation


COPPER_RESISTIVITY = 1.724 * 10**-8


# This is a single torquer
class Magnetorquer:
    def __init__(
        self,
        config,
        IdMtb,
        resistance = 3.25e-7,
        Across = 5.432e-3,
        N_turns = 64,
        max_voltage=5,
        max_current_rating=1,  # A
        max_power=0.25,  # W
    ) -> None:
        self.max_voltage = config["max_voltage"]
        self.R = config["mtb_resistance"]
        self.N_per_face = config["N_turns"]
        self.A_cross = config["A_cross"]
        self.max_power = config["max_power"]
        self.max_current_rating = config["max_current_rating"] 

        self.max_current = np.min([self.max_current_rating, np.sqrt(self.max_power / self.R) ])
        self.max_current = np.min([self.max_current, self.max_voltage / self.R ])
        self.max_voltage = self.R * self.max_current
        self.max_power   = self.R * self.max_current ** 2
        self.max_dipole_moment = self.N_per_face * self.max_current * self.A_cross  # A*m^2
        self.G_mtb_b = np.array(config["mtb_orientation"][(3*IdMtb):(3*(IdMtb+1))]).T
        self.dipole_moment = np.zeros(3,)
        self.current = 0.0
        self.voltage = 0.0
        self.power   = 0.0

    def get_dipole_moment(self):
        return self.dipole_moment


    def get_torque(self, MAG_FIELD):
        """
        Update voltage or current before getting the torque.
        """
        # TODO: Get moment and field in whatever frame the sim wants torque
        return crossproduct(self.dipole_moment) @ MAG_FIELD


    def get_power(self):
        return  self.R * self.current ** 2

    def set_voltage(
        self,
        voltage: float,
    ) -> None:
        if np.abs(voltage) > self.max_voltage:
            raise ValueError("Voltage exceeds maximum voltage rating.")
        # Current driver is PWM
        self.voltage = voltage
        self.current = voltage / self.R
        self.power   = self.R * self.current ** 2
        if abs(self.current) > self.max_current:
            raise ValueError(
                f"Current exceeds maximum power limit of {self.max_power} W."
            )
        self.dipole_moment = self.convert_current_to_dipole_moment(self.current)

    def set_current(
        self,
        current: float,
    ) -> None:
        if np.abs(current) > self.max_current:
            raise ValueError(
                f"Current exceeds maximum power limit of {self.max_power} W."
            )
        self.dipole_moment = self.convert_current_to_dipole_moment(current)

    def set_dipole_moment(
        self,
        dipole_moment: float,
    ) -> None:
        if np.abs(dipole_moment) > self.max_dipole_moment:
            raise ValueError(
                f"Dipole Moment exceeds maximum dipole moment limit of {self.max_dipole_moment} Cm."
            )
        self.dipole_moment = dipole_moment * self.G_mtb_b

    def get_dipole_moment_over_current(self) -> float:
        return self.N_per_face * self.A_cross

    def convert_current_to_dipole_moment(
        self,
        current: float,
    ) -> np.ndarray:
        #TODO: confirm that G_mtb_b is unit vector
        current = np.clip(a=current, a_min=-self.max_current, a_max=self.max_current)
        return self.N_per_face * current * self.A_cross * self.G_mtb_b

    def convert_dipole_moment_to_voltage(
        self,
        dipole_moment: np.ndarray,
    ) -> float:
        self.dipole_moment = np.clip(a=dipole_moment, a_min=-self.max_dipole_moment, a_max=self.max_dipole_moment)
        I = self.convert_dipole_moment_to_current(self.dipole_moment)
        # NOTE : HARDCODED a safety margin
        self.current = np.clip(a=I, a_min=-0.9*self.max_current, a_max=0.9*self.max_current)
        # clip voltage to max voltage
        self.voltage = np.clip(a=self.current * self.R, a_min=-self.max_voltage, a_max=self.max_voltage)
        self.power   = self.R * self.current ** 2
        return self.voltage

    def convert_voltage_to_dipole_moment(
        self,
        voltage: float,
    ) -> np.ndarray:
        self.voltage = voltage
        I = voltage / self.R
        # clip current to max current
        self.current = np.clip(a=I, a_min=-self.max_current, a_max=self.max_current)
        self.dipole_moment = self.convert_current_to_dipole_moment(self.current)
        self.power   = self.R * self.current ** 2
        return self.dipole_moment

    def convert_dipole_moment_to_current(
        self,
        dipole_moment: np.ndarray,
    ) -> float:
        dipole_moment = np.clip(a=dipole_moment, a_min=-self.max_dipole_moment, a_max=self.max_dipole_moment)
        return dipole_moment / self.N_per_face / self.A_cross

    def compute_coil_resistance(self):
        coil_length = 4 * (self.pcb_side_max - self.N*self.coil_width) \
                        * self.N * self.pcb_layers
        R =  COPPER_RESISTIVITY * coil_length \
            / (self.trace_width * self.trace_thickness)
        return R

File: actuators/Magnetorquer.cpp
#include "Magnetorquer.h"

#include "math/EigenWrapper.h"
#include <cmath>
#include <utility>
#include <iostream>

Magnetorquer::Magnetorquer(int N_MTBs, VectorXd maxVolt, VectorXd coilsPerLayer, VectorXd layers, VectorXd traceThickness,
                           VectorXd pcbSideMax, VectorXd traceWidth, VectorXd gapWidth, VectorXd maxPower, VectorXd maxCurrentRating,
                           MatrixXd mtb_orientation) : num_MTBs(N_MTBs), max_voltage(std::move(maxVolt)), N(std::move(coilsPerLayer)), pcb_layers(std::move(layers)), 
                           trace_thickness(std::move(traceThickness)), pcb_side_max(std::move(pcbSideMax)),
                           trace_width(std::move(traceWidth)), gap_width(std::move(gapWidth)), max_power(std::move(maxPower)), 
                           max_current_rating(std::move(maxCurrentRating)) 
{
    
    coil_width = trace_width + gap_width;
    
    N_per_face = VectorXd::Zero(num_MTBs);
    A_cross = VectorXd::Zero(num_MTBs);
    coil_length = VectorXd::Zero(num_MTBs);
    resistance = VectorXd::Zero(num_MTBs);
    max_dipole_moment = VectorXd::Zero(num_MTBs);
    
    for (int i=0; i<num_MTBs; i++) {
        N_per_face(i) = N(i)*pcb_layers(i);
        max_current_rating(i) = std::min(max_power(i) / max_voltage(i), max_current_rating(i));
        A_cross(i) = pow(pcb_side_max(i) - N(i)*coil_width(i), 2.0);
            
        coil_length(i) = 4*(pcb_side_max(i) - N(i)*coil_width(i))*N(i)*pcb_layers(i);
        resistance(i) = COPPER_RESISTIVITY*coil_length(i)/(trace_width(i)*trace_thickness(i));
            
        max_current_rating(i) = std::min(max_current_rating(i), sqrt(max_power(i) / resistance(i)) );
        max_current_rating(i) = std::min(max_current_rating(i), max_voltage(i) / resistance(i));
        max_voltage(i) = resistance(i) * max_current_rating(i);
        max_power(i)   = resistance(i) * pow(max_current_rating(i), 2);
        max_dipole_moment(i) = N_per_face(i) * max_current_rating(i) * A_cross(i);
    }
   
    G_mtb_b = std::move(mtb_orientation);
}

Vector3 Magnetorquer::getTorque(VectorXd voltages, Quaternion q, Vector3 magnetic_field)
{
    
    Vector3 magnetic_field_b = q.toRotationMatrix().transpose()*magnetic_field; // quaternion rotates vector in body frame to ECI. We need the inverse rotation

    Eigen::VectorXd currents = voltages.cwiseProduct(resistance.cwiseInverse());
    Eigen::VectorXd power = voltages.cwiseProduct(currents);

    Eigen::VectorXd max_current = currents.cwiseAbs();
    Eigen::VectorXd max_applied_voltage = voltages.cwiseAbs();
    Eigen::VectorXd max_applied_power = power.cwiseAbs();

    for (int i=0; i<num_MTBs; i++) {
        assert(max_current(i) <= max_current_rating(i));
        assert(max_applied_voltage(i) <= max_voltage(i));
        assert(max_applied_power(i) <= max_power(i));
    }

    //auto dipole_moments = N_per_face*A_cross*
    Vector3 dipole_moment;
    MatrixXd torque = MatrixXd::Zero(3, num_MTBs);
    for (int i=0; i<num_MTBs; i++) {
        dipole_moment = N_per_face(i)*currents(i)*A_cross(i)*G_mtb_b.col(i);
        torque.col(i) = dipole_moment.cross(magnetic_field_b);
    }

    return torque.rowwise().sum();
}